src/member.ts

```diff
@@ -0,0 +1,358 @@
+    }).length;
+  }
+
+  numberAccepted(): number {
+    return Object.keys(this.signatures).filter((id: ID) => {
+      return this.signatures[id].response === "accept";
+    }).length;
+  }
+
+  isConflicted(): boolean {
+    return Object.keys(this.signatures).some((id: ID) => {
+      return this.signatures[id].response === "conflict";
+    });
+  }
+
+  isCancelled(): boolean {
+    return this.signatures[this.from]?.response === ("cancel" as ResponseType);
+  }
+
+  finish() {
+    this.finished = true;
+  }
+}
+
+function allowAll(request: Request): boolean {
+  return true;
+}
+
+export interface ShouldAcceptCB {
+  (request: Request): boolean;
+}
+
+export interface MemberConstructorOptions {
+  id?: ID;
+  initiator?: ID;
+  shouldAccept?: ShouldAcceptCB;
+}
+
+export class Member extends EventEmitter {
+  id: ID;
+  ownFeed: FeedItem[];
+  knownFeeds: {
+    [id in ID]: FeedItem[];
+  };
+  memberIndexes: {
+    [id in ID]: number;
+  };
+  ownIndex: number;
+  requests: {
+    [id in ID]: RequestState;
+  };
+  knownMembers: string[];
+  shouldAccept: ShouldAcceptCB;
+
+  constructor(
+    { id, initiator, shouldAccept = allowAll }: MemberConstructorOptions = {
+      shouldAccept: allowAll,
+    }
+  ) {
+    super();
+    this.id = id || makeID();
+    this.ownFeed = [];
+    this.ownIndex = 0;
+    this.knownFeeds = {
+      [this.id]: this.ownFeed,
+    };
+    this.memberIndexes = {
+      [this.id]: 0,
+    };
+    this.requests = {};
+    this.knownMembers = [initiator || this.id];
+    this.shouldAccept = shouldAccept;
+  }
+
+  sync(member: Member) {
+    // Iterate through member's knownFeeds
+    // Get any new data you don't have
+    // Get new data from members to knownFeeds
+    for (const id of Object.keys(member.knownFeeds)) {
+      const feed = member.knownFeeds[id];
+      const ownCopy = this.getFeedFor(id);
+      if (ownCopy.length < feed.length) {
+        this.updateFeedFor(id, feed);
+      }
+    }
+  }
+
+  processFeeds(): boolean {
+    let hasProcessed = false;
+    let hasResponded = false;
+    for (const id of Object.keys(this.knownFeeds)) {
+      const index = this.getMemberIndexFor(id);
+      const feed = this.getFeedFor(id);
+      const item = feed[index];
+
+      if (!item) continue;
+      hasProcessed = true;
+
+      if (isRequest(item)) {
+        this.trackRequest(item);
+
+        // TODO: Detect detect concurrent requests and merge
+        if (this.isMember()) {
+          hasResponded = true;
+          const shouldAccept = this.shouldAccept(item);
+          if (shouldAccept) this.acceptRequest(item);
+          else this.denyRequest(item);
+        }
+        this.incrementMemberIndexFor(id);
+      } else if (isResponse(item)) {
+        if (this.hasRequest(item.id)) {
+          const req = this.getRequest(item.id);
+          req.addResponse(id, item);
+          let neededSignatures = this.knownMembers.length;
+          if (req.operation === "remove") neededSignatures--;
```

Comment:
Shouldn't this be always `n-1`?

Reply:
If we do `n-1` can we avoid ambiguity? Like, if we have A,B,C,D,E, if A,B,C have signed, and D/E aren't connected to each other / accept, they'll both think they're the final needed signature and cause a fork.

Reply:
If `n-1` doesn't work in that case it doesn't work in the other case, as two devices can mutually remove each other, it would also create that ambiguity.

Reply:
We had a conversation about allowing forks but maybe auto-merging them?